Update note on ptrace

mikelle-rogers opened the pull request:
## Summary

Clarified information regarding ptrace

Fixes [2246](https://github.com/dotnet/diagnostics/issues/2246)

mikelle-rogers committed:
Update note on ptrace

hoyosjs commented:
I wouldn't say this fixes 2246 - that one had a lot more nuances.

hoyosjs commented on docs/core/diagnostics/dumps.md:
```diff
@@ -13,7 +13,7 @@ A dump is a file that contains a snapshot of the process at the time it was crea
 
 > [!NOTE]
-> Collecting a dump inside a container requires PTRACE capability, which can be added via `--cap-add=SYS_PTRACE` or `--privileged`.
+> Collecting a dump requires that the process have access to ptrace targeting parent processes. This means /proc/sys/kernel/yama should have ptrace_scope set to 1 or 0. If running under any type of Open Container Initiatory technology, the seccomp profile must allow for calls to ptrace. PTRACE capability can be added via `--cap-add=SYS_PTRACE` or `--privileged`. Afterwards, the dump save path con ...
```
I'd move the path to another point here. This note is purely a point on ptrace, and the 4.8 example is still pertinent I feel. Regarding the path one, maybe the notes on the environment variable are a better place for it. These caveats need to be visible somehow from the different dump docs (i.e dotnet-dump also needs similar notices).

mikelle-rogers replied:
I don't know what the 4.8 example is referring to. 
What do you think about creating a file called NotesAboutDumps (or something else) and have every dump page have a link to that file?

hoyosjs replied:
Having a dump FAQ seems sensible to me. As for the 4.8, it refers to the kernel version running the application (docker and other container solutions block ptrace on 4.8 and below as they had security layering issues prior to that and the pressure valve they provided for people that needed it was to require CAP_SYS_PTRACE).

mikelle-rogers commented:
> I wouldn't say this fixes 2246 - that one had a lot more nuances.

I am trying to understand all of the necessary changes and plan to do them all in this PR.

mikelle-rogers commented:
@hoyosjs Where should I link this? 
https://github.com/dotnet/diagnostics/issues/2246#issuecomment-832324714
We should also cross-link to the default seccomp profile for both containerd and moby since the upstream may change (as has happened in the past 😅)

hoyosjs commented:
> @hoyosjs Where should I link this? [dotnet/diagnostics#2246 (comment)](https://github.com/dotnet/diagnostics/issues/2246#issuecomment-832324714) We should also cross-link to the default seccomp profile for both containerd and moby since the upstream may change (as has happened in the past 😅)

A good example is saying, "For example, moby+containerd specifies a [seccomp profile](https://github.com/moby/moby/blob/v20.10.18/profiles/seccomp/default.json#L408-L419) that allows parent ptrace if the container host has a kernel version higher than 4.8"

mikelle-rogers committed:
Access to the file is imperative

When writing a dump, make sure you have access to the file you write the dump to.

mikelle-rogers committed:
Remove information, link to new file.

mikelle-rogers committed:
Creating a dumps FAQ page

mikelle-rogers committed:
Where to find the FAQs of dumps.

mikelle-rogers committed:
FAQ info + write permission note to dotnet-dump

mikelle-rogers committed:
add caveats to linux dump debugging

mikelle-rogers committed:
Update dumps.md

mikelle-rogers committed:
Fix broken link

mikelle-rogers committed:
Update date, header, whitespace

mikelle-rogers committed:
Update note

mikelle-rogers committed:
Fix whitespace

mikelle-rogers committed:
Update permission information

mikelle-rogers committed:
update dumps to be more general

mikelle-rogers committed:
Update permissions and caveats comments

mikelle-rogers committed:
Update linux dumps to include caveats

mikelle-rogers committed:
Add caveats in list

mikelle-rogers commented:
@hoyosjs Do you know how to fix this link? Line 15: [Warning] Invalid file link: '../install/macos-notarization-issues'.

hoyosjs commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+- For applications running under an Open Container Initiative technology, the `seccomp` profile must allow for calls to `ptrace`. For example, moby+containerd specifies a default [seccomp profile](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) that allows `ptrace` only if the container host has a kernel version higher than 4.8 or if the `CAP_SYS_PTRACE` capability was spe ...
+- on macOS, the use of `ptrace` requires the host of the target process to be properly entitled. See [Default Entitlements](../install/macos-notarization-issues#default-entitlements) for more information.
+
+## See also
```
Not sure if see also makes sense in a FAQ, but don't feel strongly about it either

hoyosjs commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+
+- On Linux-based systems, `/proc/sys/kernel` should have `ptrace_scope` set to `1` or `0`. Any value higher than this will require running as root or will block dump collection entirely.
+- For applications running under an Open Container Initiative technology, the `seccomp` profile must allow for calls to `ptrace`. For example, moby+containerd specifies a default [seccomp profile](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) that allows `ptrace` only if the container host has a kernel version higher than 4.8 or if the `CAP_SYS_PTRACE` capability was spe ...
+- on macOS, the use of `ptrace` requires the host of the target process to be properly entitled. See [Default Entitlements](../install/macos-notarization-issues#default-entitlements) for more information.
```
```suggestion
- on macOS, the use of `ptrace` requires the host of the target process to be properly entitled. See [Default Entitlements](../install/macos-notarization-issues.md#default-entitlements) for more information.
```

gewarren replied:
```suggestion
- On macOS, the use of `ptrace` requires the host of the target process to be properly entitled. See [Default Entitlements](../install/macos-notarization-issues.md#default-entitlements) for more information.
```

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+---
```
This new file will need an entry in the table of contents.

gewarren replied:
Also, you might want to convert this to a YAML file to optimize the questions and answers for a search engine: https://review.learn.microsoft.com/en-us/help/contribute/contribute-how-to-faq-guide?branch=main.

gewarren replied:
Finally, all Markdown files need to be named using lower-case characters only.

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+---
+title: Dumps - .NET FAQ
```
```suggestion
title: Dumps FAQ - .NET
```

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+---
+title: Dumps - .NET FAQ
+description: Frequently Asked Questions about dumps .NET
```
```suggestion
description: Get answers to frequently asked questions about dumps in .NET.
```

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+ms.date: 09/19/2022
+---
+
+# Frequently Asked Questions Regarding Dumps
```
```suggestion
# Dumps: Frequently asked questions
```

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+
+# Frequently Asked Questions Regarding Dumps
+
+## Collecting dumps on macOS and Linux
```
```suggestion
## How do I collect dumps on macOS and Linux?
```

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+
+Collecting a dump requires that the processes in the system have access to call `ptrace` targeting parent processes. The following is a non-exhaustive list of settings that are needed for this to work properly:
+
+- On Linux-based systems, `/proc/sys/kernel` should have `ptrace_scope` set to `1` or `0`. Any value higher than this will require running as root or will block dump collection entirely.
```
```suggestion
- On Linux-based systems, set `ptrace_scope` for `/proc/sys/kernel` to `1` or `0`. Any value higher than this requires running as root to prevent dump collection from being blocked.
```

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+Collecting a dump requires that the processes in the system have access to call `ptrace` targeting parent processes. The following is a non-exhaustive list of settings that are needed for this to work properly:
+
+- On Linux-based systems, `/proc/sys/kernel` should have `ptrace_scope` set to `1` or `0`. Any value higher than this will require running as root or will block dump collection entirely.
+- For applications running under an Open Container Initiative technology, the `seccomp` profile must allow for calls to `ptrace`. For example, moby+containerd specifies a default [seccomp profile](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) that allows `ptrace` only if the container host has a kernel version higher than 4.8 or if the `CAP_SYS_PTRACE` capability was spe ...
```
```suggestion
- For applications running under an Open Container Initiative technology, the `seccomp` profile must allow for calls to `ptrace`. For example, [containerd](https://github.com/moby/containerd) specifies a default [seccomp profile](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) that allows `ptrace` only if the container host has a kernel version higher than 4.8 or if the `CAP_SYS_PTRACE` capability was specified.
```

gewarren replied:
Is the moby part needed? It just looked a little weird to me.

hoyosjs replied:
Moby builds on containerd. Moby passes a profile to containerd, so it's a compound tech. I can think of a better way of phrasing this.

hoyosjs replied:
```suggestion
- For applications running under an Open Container Initiative technology, the `seccomp` profile must allow for calls to `ptrace`. For example, `Docker` uses [containerd](https://github.com/moby/containerd) under the hood as a container runtime. When initializing the runtime, it specified specifies a default [seccomp profile](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) that allows `ptrace` only if the container host has a kernel version higher than 4.8 or if the `CAP_SYS_PTRACE` capability was specified.
```

hoyosjs replied:
@gewarren does this look ok without needing to go too much into detail in extraneous techologies?

gewarren replied:
Looks great thanks - I'm just going to make one little edit to remove a duplicated word.

gewarren commented on docs/core/diagnostics/FAQ-dumps.md:
```diff
@@ -0,0 +1,23 @@
+
+## See also
+
+Learn more about how you can leverage dumps to help diagnosing problems in your .NET application.
```
```suggestion
Learn more about how you can leverage dumps to help diagnose problems in your .NET application.
```

gewarren commented on docs/core/diagnostics/debug-linux-dumps.md:
```diff
@@ -10,6 +10,8 @@ ms.date: 08/27/2020
 
 ## Collect dumps on Linux
 
+[!NOTE] Frequently asked questions regarding dump collection, analysis and other caveats can be found [here](FAQ-dumps.md).
```
```suggestion
> [!TIP]
> For frequently asked questions about dump collection, analysis, and other caveats, see [Dumps: FAQ](faq-dumps.md).
```

gewarren commented on docs/core/diagnostics/dotnet-dump.md:
```diff
@@ -11,6 +11,9 @@ ms.topic: reference
 > [!NOTE]
 > `dotnet-dump` for macOS is only supported with .NET 5 and later versions.
 
+> [!NOTE]
```
```suggestion
> [!TIP]
```

gewarren commented on docs/core/diagnostics/dotnet-dump.md:
```diff
@@ -100,7 +103,7 @@ dotnet-dump collect [-h|--help] [-p|--process-id] [-n|--name] [--type] [-o|--out
 - **`-o|--output <output_dump_path>`**
 
-  The full path and file name where the collected dump should be written.
+  The full path and file name where the collected dump should be written. Ensure the user under which the dotnet process is running has write permissions to the directory that was specified.
```
```suggestion
  The full path and file name where the collected dump should be written. Ensure that the user under which the dotnet process is running has write permissions to the specified directory.
```

gewarren commented on docs/core/diagnostics/dumps.md:
```diff
@@ -12,11 +12,12 @@ A dump is a file that contains a snapshot of the process at the time it was crea
 > [!NOTE]
 > Dumps may contain sensitive information because they can contain the full memory of the running process. Handle them with any security restrictions and guidances in mind.
 
+> [!NOTE]
```
```suggestion
> [!TIP]
```

gewarren commented on docs/core/diagnostics/dumps.md:
```diff
@@ -12,11 +12,12 @@ A dump is a file that contains a snapshot of the process at the time it was crea
 > Dumps may contain sensitive information because they can contain the full memory of the running process. Handle them with any security restrictions and guidances in mind.
 
+> [!NOTE]
+> Frequently asked questions regarding dump collection, analysis and other caveats can be found [here](FAQ-dumps.md).
```
```suggestion
> For frequently asked questions about dump collection, analysis, and other caveats, see [Dumps: FAQ](faq-dumps.md).
```

gewarren commented on docs/core/diagnostics/dumps.md:
```diff
@@ -27,7 +28,7 @@ The following table shows the environment variables you can configure for collec
 |`COMPlus_DbgEnableMiniDump` or `DOTNET_DbgEnableMiniDump`|If set to 1, enable core dump generation.|0|
 |`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|
-|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to.|`/tmp/coredump.<pid>`|
+|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to. Ensure the user under which the dotnet process is running has write permissions to the directory specified.|`/tmp/coredump.<pid>`|
```
```suggestion
|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to. Ensure that the user under which the dotnet process is running has write permissions to the specified directory.|`/tmp/coredump.<pid>`|
```

gewarren commented on docs/core/diagnostics/dumps.md:
```diff
@@ -27,7 +28,7 @@ The following table shows the environment variables you can configure for collec
-|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to.|`/tmp/coredump.<pid>`|
+|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to. Ensure the user under which the dotnet process is running has write permissions to the directory specified.|`/tmp/coredump.<pid>`|
 |`COMPlus_CreateDumpDiagnostics` or `DOTNET_CreateDumpDiagnostics`|If set to 1, enable diagnostic logging of dump process.|0|
 |`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later) If set to 1, the runtime generates a JSON-formatted crash report thta includes information about the threads and stack frames of the crashing application. The crash report name is the dump path/name with *.crashreport.json* appended.
```
```suggestion
|`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later) If set to 1, the runtime generates a JSON-formatted crash report that includes information about the threads and stack frames of the crashing application. The crash report name is the dump path or name with *.crashreport.json* appended.
```

gewarren commented on docs/core/diagnostics/dumps.md:
```diff
@@ -27,7 +28,7 @@ The following table shows the environment variables you can configure for collec
 |`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|
-|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to.|`/tmp/coredump.<pid>`|
+|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to. Ensure the user under which the dotnet process is running has write permissions to the directory specified.|`/tmp/coredump.<pid>`|
 |`COMPlus_CreateDumpDiagnostics` or `DOTNET_CreateDumpDiagnostics`|If set to 1, enable diagnostic logging of dump process.|0|
```
```suggestion
|`COMPlus_CreateDumpDiagnostics` or `DOTNET_CreateDumpDiagnostics`|If set to 1, enables diagnostic logging of dump process.|0|
```

mikelle-rogers committed:
Response to review suggestions

Created an FAQ yaml page to optimize for searching.

mikelle-rogers committed:
Change from note to tip

mikelle-rogers committed:
Clarify dotnet-dump

mikelle-rogers committed:
Update specific details.

mikelle-rogers committed:
Correct sections

mikelle-rogers committed:
Move text to the second line of TIP

mikelle-rogers committed:
Update toc.yml

gewarren commented on docs/core/diagnostics/debug-linux-dumps.md:
```diff
@@ -10,6 +10,9 @@ ms.date: 08/27/2020
 
 ## Collect dumps on Linux
 
+> [!TIP] 
```
```suggestion
> [!TIP]
```

gewarren commented on docs/core/diagnostics/debug-linux-dumps.md:
```diff
@@ -10,6 +10,9 @@ ms.date: 08/27/2020
 ## Collect dumps on Linux
 
+> [!TIP] 
+> For frequently asked questions about dump collection, analysis and other caveats, see [DUMPS: FAQ](faq-dumps.yml).
```
```suggestion
> For frequently asked questions about dump collection, analysis, and other caveats, see [Dumps: FAQ](faq-dumps.yml).
```

gewarren commented on docs/core/diagnostics/dotnet-dump.md:
```diff
@@ -11,6 +11,9 @@ ms.topic: reference
 > `dotnet-dump` for macOS is only supported with .NET 5 and later versions.
 
+> [!TIP]
+> Frequently asked questions regarding dump collection, analysis and other caveats can be found [here](faq-dumps.yml).
```
```suggestion
> For frequently asked questions about dump collection, analysis, and other caveats, see [Dumps: FAQ](faq-dumps.yml).
```

gewarren commented on docs/core/diagnostics/faq-dumps.yml:
```diff
@@ -0,0 +1,31 @@
+      - question: Where can I learn more about how I can leverage dumps to help diagnose problems in my .NET application?
+        answer: | 
+          Here are some additional resources: 
+          - [How do I debug a linux dump?](debug-linux-dumps.md) 
```
```suggestion
          - [Debug Linux dumps](debug-linux-dumps.md) 
```

gewarren commented on docs/core/diagnostics/faq-dumps.yml:
```diff
@@ -0,0 +1,31 @@
+        answer: | 
+          Here are some additional resources: 
+          - [How do I debug a linux dump?](debug-linux-dumps.md) 
+          - [How do I debug a deadlock in my .NET application?](debug-deadlock.md) 
```
```suggestion
          - [Debug a deadlock in a .NET app](debug-deadlock.md) 
```

mikelle-rogers committed:
Update docs/core/diagnostics/debug-linux-dumps.md

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

mikelle-rogers committed:
Update docs/core/diagnostics/debug-linux-dumps.md

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

mikelle-rogers committed:
Update docs/core/diagnostics/dotnet-dump.md

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

mikelle-rogers committed:
Update docs/core/diagnostics/faq-dumps.yml

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

mikelle-rogers committed:
Update docs/core/diagnostics/faq-dumps.yml

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

gewarren committed:
Update faq-dumps.yml

gewarren committed:
Update faq-dumps.yml

gewarren committed:
Try single quotes

gewarren committed:
Try something else

gewarren committed:
Keep trying

gewarren committed:
Run it through a validator
